오픈마켓2 [Step1] Kyo, LJ

KyoPak opened the pull request:
@Monsteel
안녕하세요 토니!
Kyo(@KyoPak)와 LJ(@lj-7-77)가 OpenMarket2 Step1의 구현 후 PR 작성해보았습니다! 
생각보다 구현내용이 많아 PR이 많이 늦어졌습니다..!
상품 수정화면과 Patch, Delete 메서드들은 현재 임의의 데이터를 사용하고 있으며 Step2를 진행하며 수정하겠습니다.!
토니의 좋은 코멘트 부탁드립니다 :) 

## ✓ 진행단계 : Step-1

## 📝 STEP 진행 중 경험하고 배운 것
 
- CollectionView
    - ✅ Collection View의 활용
    - ✅ Cell 내부 Button Action Delegate 처리 
- Network
    - ✅ multipart/form-data의 구조 파악
    - ✅ URLSession을 활용한 multipart/form-data 요청 전송
    - ✅ NetworkManager POST, PATCH, DELETE Method 구현
    
- 추가 구현 
    - ✅ Pagination 구현
    - ✅ PHPickerView 구현
    - ✅ 사용자 친화적인 UI/UX 구현 (적절한 입력 컴포넌트 사용, 알맞은 키보드 타입 지정)
    - ✅ 키보드와 View의 간섭 해결
    - ✅ 상속 혹은 프로토콜 기본구현을 통해 (수정/등록 과정의) 공통기능 구현

---
## ⚒️ 코드 주요 구현내용 

<details>
<summary>Details</summary>
<div markdown="1">

- RootProductViewController
    - 상품등록과 상품수정의 공통기능(NavigationItem-cancel버튼,done버튼)을 가진 뷰컨트롤러입니다. 
    - `AddViewController`와 `ModifyViewController`가 이 클래스를 상속받아 버튼을 눌렀을 때 alert각각 역할에 맞는 기능을 수행하도록 구현하였습니다.
    - `showView` 전역변수를 선언하여 `AddViewController`에서는 `AddProductView`를, `ModifyViewController`에서는 `ModifyProductView`를 보여주도록 구현하였습니다
    
- AddViewController
    - 상품등록 시 이미지를 첨부하는 기능을 PHPickerView를 활용하여 구현하였습니다.
    - `ImageCollectionViewCellDelegate` 프로토콜을 채택하여 등록가능한 최대 이미지개수와 asset타입(이미지, 라이브포토, 비디오)을 이미지만 가능하도록 설정하였습니다. 

- NewProduct
상품등록을 위해 POST 할 때 사용하기 위한 타입입니다. 

- NetworkManager - `postData(url:newData:completion:)`, `patchData(url:)`, `fetchDeleteDataURI(to: url)`, `deleteProduct(url:)`
    - `postData(url:newData:completion:)` : 상품등록할 때 http 메세지를 작성하고, 등록할 상품데이터를 보내주는 메서드입니다.
    - patchData(url:)` : 상품수정 시 수정할 데이터를 encoding하여 보내주는 메서드입니다.
    - `fetchDeleteDataURI(to: url)` : 삭제할 상품의 URI 응답값을 받아오는 메서드입니다.
    - `deleteProduct(url:)` : 삭제할 상품의 URI를 받아온 후 해당 URI를 삭제요청하는 메서드입니다.
    
- UIImage+Extension 
    - 이미지 리사이징을 위한 메서드들을 확장하여 구현하였습니다.
    - `compressTo(expectedSizeInKb:)` : compressionQuality를 줄여나가는 메서드입니다.
    - `resize(expectedSizeInKb:)` : width, height를 줄여나가는 메서드입니다.
    
</div>
</details>

---
## 🤔 고민 

### 1️⃣ 페이지네이션 구현시 지나치게 `setupData()` 를 여러번 호출하는 문제
- `scrollViewDidScroll` 내부에서 스크롤이 일정부분 내려간다면 pageNo를 늘려 새로운 데이터를 추가로 가져오게끔 구현하였습니다.
- 이 과정에서 빠르게 스크롤할 경우 `scrollViewDidScroll`가 매우 많이 호출되는 것을 확인할 수 있었습니다.
- 때문에 내부의 `setupData()` (데이터를 가져오는 메서드) 또한 많이 호출되었습니다.
- scrollState를 만들어 만약 로딩중이라면 중복으로 setupData()를 호출하지 못하게끔 구현하여 빠른 스크롤하는 경우의 메서드호출을 방지하였습니다.
```swift
 private func setupData() {
    defer { pageCount += Constant.pageNumberUnit.rawValue }
        
    guard scrollState == .idle else { return }
    scrollState = .isLoading
    ...    
    networkManager.fetchData(to: productListURL, dataType: ProductPage.self) { result in
        switch result {
        case .success(let data):
            ...
            ...
            self.scrollState = .idle
            }
        case .failure(let error):
            ...
        }
    }
}
```

### 2️⃣ 프로토콜 vs 상속 

- 상품수정화면과 상품등록화면의 View와 Controller가 유사하게 구현될 필요가 있었습니다.
- 때문에 프로토콜과 상속을 고민하였고 상속을 선택하였습니다.
- 상속을 선택한 이유는 아래와 같습니다.
- 1. 일단 메서드가 겹치는 경우가 많았습니다. 때문에 상속을 사용하는 방향이 프로토콜의 기본구현을 사용하는 것보다 효율적이라고 생각했습니다.
- 2. 겹치는 메서드 내부에서 프로퍼티에 접근하는 경우가 있었습니다.
- 만약 메서드가 서로 불필요한 메서드가 있거나 굳이 상속받아야할 필요가 없다면 프로토콜을 고려했을 것이지만 
- 상품수정화면과 상품등록화면은 View는 동일하였고 Controller 부분도 `PHPicker` 구현부를 제외하면 유사하였습니다.


---
## ❓ 조언을 얻고 싶은 부분
### 1️⃣ Image Resizing, Resizing 소요 시간

- 높은 용량의 이미지를 Resizing 한 첫번째 방법은 `compressionQuality`를 0.1씩 낮춰나가는 것이었습니다.
- 하지만 해당 방법은 매우 높은 용량의 이미지의 경우 `compressionQuality`가 0.1이어도 원하는 용량보다 훨씬 컸습니다.

- 두번째 방법은 해당 이미지의 width를 원하는 용량보다 작을때까지 1/10으로 줄이는 방안이었습니다.
- 하지만 해당방법은 `self.draw` 메서드가 여러번 호출되어야했고 `draw()`는 view를 다시 그리는 메서드이기 때문에 상당히 소요시간이 많이 걸렸습니다.
- 그리고 시간이 오래걸리는 `draw()` 메서드가 가급적 적게 호출해야하기때문에 1/10로 감소율을 선택하게 되었고, 이미지 손실률이 매우 클 것이라고 생각이 들었습니다.

- 토니였다면 어떤 방법으로 용량을 줄여나갈지 궁금합니다....!
 

### 2️⃣ PHPicker에 대한 이슈

- 일단 UIImagePickerController가 부분적으로 decrecated가 되기때문에 PHPicker을 사용해보았습니다.
- PHPicker는 UIImagePickerControll에서 사용할 수 없는 이미지나 동영상 기능 추가할수가 있었고,
- multiselect / zoom in or out / search를 지원하고 권한요청 팝업이 필요없다는게 좋은점이라고 생각들었습니다.
- 하지만 아래의 두가지 이슈사항이 있었고 현재 해결에 힘쓰고있습니다.

#### 1. 앨범내의 첫번째 이미지만 Pick이 되지 않는 문제
- Xcode 시뮬레이터의 앨범에서 이미지를 Pick하여 상품등록 이미지에 첨부하려 할 때, 이미지가 nil로 불러와지는 문제가 있습니다.
- 해당문제는 코드문제가 아니라 다른 확장자로 인한 시뮬레이터 버그로 보여지는데 혹시 토니가 알고 계신 부분이 있으실까요..?!
참고링크 : https://github.com/ionic-team/capacitor-plugins/issues/587

#### 2. PHPickViewController의 allowEdit 지원 문제 
- 기존의 UIPicker는 allowsEditing라는 크롭기능을 사용할 수 있는 instance 프로퍼티가 존재했지만 PHPicker에서는 이러한 기능을 찾을 수 없었습니다.
- 혹시 토니가 알고 계시다면 조언 부탁드립니다...!

KyoPak committed:
feat : Pagination 구현 및 Constant enum 생성

- NetworkRequest 내부 productList 연관값으로 변경
- scrollViewDidScroll() 메서드 내부 구현
- Constant enum 생성하여 내부에 상수 보관 처리
- ScrollState enum 생성하여 Scroll 상태 플래그 처리

limjy0 committed:
refactor : Product 모델 일부 타입 변경 및 dateDecodingStrategy 활용

- Product 모델 createdAt, issuedAt 타입 String ➡️ Date 로 변경
- dateDecodingStrategy 활용

KyoPak committed:
feat : NetworkManager 내부 네트워킹 POST, PATCH, DELETE 메서드 구현

- 네트워킹 POST, PATCH, DELETE 메서드 구현
- 문자열을 데이터로 바꾸고 append해주는 appendStringData() 메서드 구현
- multipart/data-form 양식 구현시 사용되는 convertDataForm(), convertFileDataForm()메서드 구현

limjy0 committed:
refactor : 중복 데이터로드 방지를 위한 pagination 수정 및 열거형 접근제어 추가

- 중복 데이터로드 방지를 위한 pagination 수정
- 열거형 Constant, ScrollState에 접근제어 추가

KyoPak committed:
feat : NetworkRequest에 POST, PATCH, DELETE Case 추가 구현

- NetworkRequest에 POST, PATCH, DELETE Case의 URL 추가 구현
- NetworkManager 내부 URL Test용 하드코딩을 NetworkRequest 적용으로 변경

limjy0 committed:
feat : AddProductView UI 생성

- AddProductView UI Component 구현
- 오토레이아웃 구현

KyoPak committed:
feat : 상품등록 화면 내부 오토레이아웃 초기 구현 및 컬렉션 뷰 구현

- 상품등록 화면 내부 오토레이아웃 초기 구현
- 상품등록 화면 내부 컬렉션 뷰 구현 및 Cell 구현
- 페이지네이션 데이터 추가되는 부분 수정

limjy0 committed:
feat : 상품등록할 타입 NewProduct 생성 및 상품등록화면 버튼 구현

- 상품등록할 타입 NewProduct 생성
- 상품등록화면 내비게이션바 버튼 cancelButtonItem, doneButtonItem 구현

KyoPak committed:
feat : 상품등록화면에서 ImagePicker 구현 및 CollectionView Cell에 Image 바인딩 구현

- 상품등록화면에서 ImagePicker 구현
- 이미지 추가 관련 ImageCollectionViewCellDelegate 구현
- ImageCollectionViewCell에 Image 바인딩 구현

limjy0 committed:
feat : DataError 열거형 생성, doneButtonTapped(), setupData() 내부 구현

- DataError 열거형 생성
- doneButtonTapped() 메서드 내부 구현
- AddProductView에서 POST데이터를 만들어주는 setupData() 메서드 내부 구현

KyoPak committed:
feat : PostData 내부 로직 수정 및 구현

- AddViewController클래스 PostData 내부 로직 수정
- MainViewController클래스 viewWillAppear() 추가
- ViewController 내부 접근제어 추가

KyoPak committed:
docx : ReadME Update  추후 보완 예정

KyoPak committed:
fix : MainViewController 내부 Pagination  로직 변경

- pageNo  추가 되는 로직 위치 변경

KyoPak committed:
Merge remote-tracking branch 'origin/step3' into step3

limjy0 committed:
feat : UITextFieldDelegate 메서드 구현, 각 TextField별 Delegate지정 및 입력정보에 따른 키보드 설정

- UITextFieldDelegate 메서드 구현
- 각 TextField별 Delegate지정
- 입력정보에 따른 키보드 설정

KyoPak committed:
fix : 불필요한 코드 삭제 및 Pagination pageCount 코드 위치 변경

KyoPak committed:
feat : 상품 등록 성공 및 에러 발생시 보여주는 Alert 기능 구현

- UIViewController Extension에 showAlert()메서드 구현

KyoPak committed:
refactor :  RootProductView, RootProductViewController 생성 후 상속으로 분리

- RootProductView, RootProductViewController 구현
- ModifyProductView, ModifyViewController 구현
- AddProduct, ModifyProduct View / Controller 상속으로 분리

KyoPak committed:
feat : UIImage Extension 하여 compressTo(), resize() 구현

- Image Resizing 방식이 다른 2개의 Resizing 메서드 구현

limjy0 committed:
feat : 텍스트 입력 중 키보드가 콘텐츠를 가리지 않는 기능 추가

- 텍스트 입력 중 키보드가 콘텐츠를 가리지 않는 기능  추가
- setupNotification(), keyboardWillShow(notification:), keyboardWillHide() 메서드 구현

KyoPak committed:
style : 코드 컨벤션 정리 및 메서드 위치 조정

Monsteel commented on OpenMarket/OpenMarket/Controllers/AddViewController.swift:
```diff
@@ -6,12 +6,96 @@
-final class AddViewController: UIViewController {
-    let addProductView = AddProductView()
+class AddViewController: RootProductViewController {
+    var addProductView = AddProductView()
```
`final` 키워드를 제거하신 이유가 있을까요~?

KyoPak replied:
깜빡했습니다..바로 추가했습니다!

Monsteel commented on OpenMarket/OpenMarket/Controllers/RootProductViewController.swift:
```diff
@@ -0,0 +1,70 @@
+
+import UIKit
+
+class RootProductViewController: UIViewController {
```
개인적으로 `Root~~Controller`, `Base~~Controller`라는 네이밍은 지양하고 있습니다.

[리스코프 치환 원칙](https://ko.wikipedia.org/wiki/%EB%A6%AC%EC%8A%A4%EC%BD%94%ED%94%84_%EC%B9%98%ED%99%98_%EC%9B%90%EC%B9%99)중 `is a kind of 관계`를 검증하는데에 방해가 되기 때문이에요.

예를들어, 아래의 말들을 생각해볼 수 있겠습니다.
AddViewController는 RootProductViewController 이다.
AddViewController는 ProductViewController 이다.
AddViewController는 UIViewController 이다.
AddViewController는 BaseViewController 이다.

jylim72 replied:
아하 AddViewController가 RootViewController로 해석이 될 여지가 있기 때문이군요. 감사합니다. ProductViewController로 수정하였습니다. 
마찬가지로 `RootProductView` ➡️ `ProductView`로 수정하였습니다.🙇‍♀️🙇‍♂️
[a4c5805](https://github.com/yagom-academy/ios-open-market/pull/231/commits/a4c5805a05e8685d9fdf0d31d3f9eaf2dd0a5f71)

Monsteel commented on OpenMarket/OpenMarket/Controllers/AddViewController.swift:
```diff
@@ -6,12 +6,96 @@
+            return result.itemProvider
+        }
+        
+        for itemProvider in itemProvides {
```
Depth가 깊어지는 경우 function으로 분리하여 표현하면 가독에 도움이 될 수 있어요 :)

KyoPak replied:
넵 감사합니다 : )
기존의 PHPicker사용에서 UIImagePicker 사용으로 변경하여서 해당 부분 구조가 살짝바뀌었습니다!

Monsteel commented on OpenMarket/OpenMarket/Controllers/AddViewController.swift:
```diff
@@ -6,12 +6,96 @@
+// MARK: - Extension UICollectionView
+extension AddViewController: UICollectionViewDelegate, UICollectionViewDataSource {
+    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
+        return cellImages.count < 5 ? cellImages.count + 1 : 5
```
각 숫자들이 어떤 의미를 가지는지 한번에 파악하기가 어려워요 😭 좋은 방법 없을까요 ?

jylim72 replied:
넵, 이미지를 최대 5장까지만 추가할 수 있다는 의미로 숫자 5를 상수 maxImageNumber로 수정하였습니다.☺️
[a4c5805](https://github.com/yagom-academy/ios-open-market/pull/231/commits/a4c5805a05e8685d9fdf0d31d3f9eaf2dd0a5f71)

Monsteel commented on OpenMarket/OpenMarket/Controllers/MainViewController.swift:
```diff
@@ -7,35 +7,66 @@
+        case itemsPerPage = 10
+    }
+    
+    private enum ScrollState {
```
👍

Monsteel commented on OpenMarket/OpenMarket/Controllers/MainViewController.swift:
```diff
@@ -79,7 +110,16 @@ extension MainViewController {
 
 // MARK: - Extension UICollectionView
 extension MainViewController: UICollectionViewDelegate {
```
`UICollectionViewDelegate` 에 대한 부분을 따로 분리해주셨군요 👍 
이렇게 되면 function들의 관심사가 분리가 되어 가독에 도움이 되어 저도 선호합니다. 아주 좋습니다!

Monsteel commented on OpenMarket/OpenMarket/Views/AddProductView.swift:
```diff
@@ -7,14 +7,12 @@
 import UIKit
 
-final class AddProductView: UIView {
+final class AddProductView: RootProductView {
```
`RootProductView`도 앞서 리뷰한 `RootProductViewController`네이밍 리뷰와 동일한 기조로 바라볼 수 있을것 같아요 👍

Monsteel commented on OpenMarket/OpenMarket/Views/RootProductView.swift:
```diff
@@ -0,0 +1,262 @@
+    func textView(_ textView: UITextView,
+                  shouldChangeTextIn range: NSRange,
+                  replacementText text: String) -> Bool {
+        if descriptionTextView.text.count >= 1000 && text != "" {
```
이곳에서의 1000도 어떤 숫자인지 바로 파악하기 어렵습니다 😭

jylim72 replied:
넵 1000 이라는 숫자도 최대 텍스트길이를 의미하는 `maxTextCount` 상수로 선언하였습니다!
[a4c5805](https://github.com/yagom-academy/ios-open-market/pull/231/commits/a4c5805a05e8685d9fdf0d31d3f9eaf2dd0a5f71)

KyoPak replied:
해당 부분도 수정완료하였습니다!
[a4c5805](https://github.com/yagom-academy/ios-open-market/pull/231/commits/a4c5805a05e8685d9fdf0d31d3f9eaf2dd0a5f71)

Monsteel reviewed (changes requested):
안녕하세요 Kyo LJ 👍 
오픈마켓2 첫 스텝 수고 많으셨습니다! :)

조언을 얻고 싶은 부분에 대해 답변을 드려야 하는데,
이미지 리사이징이나 PHPicker를 직접적으로 사용하고 있지 않다보니 깊은 답변을 드리기에는 어려움이 있을것 같아요. 🙏 사전에 양해를 구합니다.

답변 남겨볼게요!

1. 제가 기술적 해결책을 답변해 드리기에는 어려움이 있을 것 같아요.
당장 드는 생각으로는, Image resizing 구현이 포함된 open source의 코드를 참고해 보는건 어떨까 싶은데요, 유사한 기술적 고민의 흔적을 찾을 수 있을 수 있어, 저도 즐겨 선택하고 있습니다!

2-1. 혹시 실제 디바이스에도 동일한 문제가 발생하나요? 시뮬레이터의 버그라면 크게 신경쓸 부분이 아니라고 생각되긴 하는데, 실제 디바이스에서도 발생하면 조금 깊게 들여야 보아야 할 것 같아요

2-2. PHPickViewController에 edit.. 기존 UIImagePickerController를 대체할 수 있는 다른 기능이 아직도 없어 보입니다.. 예전에 저도 비슷한 고민을 했던 기억이 나는데 결국 UIImagePickerController를 사용하는 방향이 제일 수월 (?) 해 보여서 그 방향으로 결정했던 기억이 나네요.. 💦💦

이번 스텝도 수고 많으셨습니니다 👍 
혹시나 함께 고민해보고 싶은 부분이나 추가로 궁금하신 부분이 있다면 언제든 말씀 부탁드려요 👍 

수고하셨습니다!

KyoPak committed:
refactor : PHPickerView 방식에서 UIImagePicker 방식으로 변경

- 이미지 편집 기능 사용을 위해 UIImagePicker로 기본 구현 변경
- Add, ModifyViewController final 키워드 추가

limjy0 committed:
refactor : ViewController, View 네이밍 수정, maxImageNumber, maxTextCount, alert메세지 상수 선언

- 네이밍 수정(RootProductViewController ➡️ ProductViewController)
- 네이밍 수정(RootProductView  ➡️ ProductView)
- 최대 이미지개수 maxImageNumber 상수 선언
- 최대 텍스트길이 maxTextCount 상수 선언
- alert 메세지 Constant 열거형으로 상수 선언

KyoPak committed:
refactor: resize(expectedSizeInKb:)메서드 로직 변경

- 내부 로직 단순화 및 비율 3배수 리사이징으로 변경

limjy0 committed:
fix : 상품등록 시 할인가격 표시되도록 수정

- NewProduct CodingKey 프로토콜 채택

KyoPak committed:
refactor : ProductViewController내부 showView프로퍼티 연산프로퍼티로 변경

- AddViewController, ModifyViewController 내부 showView 프로퍼티 get으로 재정의

Monsteel reviewed (approved):
LGTM! 👍
